test/unit_tests/container/container.cpp

```diff
@@ -31,11 +67,183 @@ struct handle_interface: handle_base<Identifier> {
+              c.array_dimension(field_index));
+    }
+
+    CHECK(data.template get_array_dims<field::DOff>(0) == 1ul);
+    CHECK(data.template get_array_dims<field::DOn>(0) == 1ul);
+}
+
+TEST_CASE("soa::get_num_variables", "[Neuron][data_structures]") {
+    storage data;
+
+    data.set_field_status<field::DOn>(true);
+    data.set_field_status<field::DOff>(false);
+
+    auto c = field::C{};
+
+    CHECK(data.get_num_variables<field::C>() == c.num_variables());
+    CHECK(data.get_num_variables<field::DOff>() == 1ul);
+    CHECK(data.get_num_variables<field::DOn>() == 1ul);
+}
+
+TEST_CASE("Identifier defer delete ", "[Neuron][internal][data_structures]") {
+    storage data;
+
+    REQUIRE(detail::identifier_defer_delete_storage != nullptr);
+    auto usage_before = detail::compute_identifier_defer_delete_storage_size();
+    { owning_handle instance{data}; }
+    auto usage_after = detail::compute_identifier_defer_delete_storage_size();
+
+    CHECK(usage_after.size - usage_before.size > 0);
+    CHECK(usage_after.capacity > 0);
+    CHECK(usage_before.size <= usage_before.capacity);
+    CHECK(usage_after.size <= usage_after.capacity);
+}
+
+TEST_CASE("defer delete storage pointer", "[Neuron][internal][data_structures]") {
+    REQUIRE(detail::defer_delete_storage != nullptr);
+
+    auto usage_before = detail::compute_defer_delete_storage_size();
+    { storage data; }
+    auto usage_after = detail::compute_defer_delete_storage_size();
+
+    CHECK(usage_after.size - usage_before.size > 0);
+    CHECK(usage_after.capacity > 0);
+    CHECK(usage_before.size <= usage_before.capacity);
+    CHECK(usage_after.size <= usage_after.capacity);
+}
+
+template <class Tag, class Storage>
+std::size_t compute_row_size(const Storage& data) {
+    std::size_t local_size = 0ul;
+    auto tag = data.template get_tag<Tag>();
+    for (int field_index = 0; field_index < detail::get_num_variables(tag); ++field_index) {
+        local_size += data.template get_array_dims<Tag>()[field_index] * sizeof(typename Tag::type);
+    }
+
+    return local_size;
+}
+
+TEST_CASE("container memory usage", "[Neuron][internal][data_structures]") {
+    storage data;
+    data.set_field_status<field::DOn>(true);
+    data.set_field_status<field::DOff>(false);
+
+    std::size_t row_size = compute_row_size<field::A>(data) + compute_row_size<field::B>(data) +
+                           compute_row_size<field::C>(data) + compute_row_size<field::DOn>(data);
+
+    auto r1 = owning_handle{data};
+    auto r2 = owning_handle{data};
+    auto r3 = owning_handle{data};
+
+    auto n_rows = data.size();
+
+    auto usage = memory_usage(data);
+
+    CHECK(usage.heavy_data.size == row_size * n_rows);
+    CHECK(usage.heavy_data.size <= usage.heavy_data.capacity);
+
+    CHECK(usage.stable_identifiers.size % n_rows == 0);
+    CHECK(usage.stable_identifiers.size >= n_rows * sizeof(std::size_t*));
+    CHECK(usage.stable_identifiers.size < n_rows * 4 * sizeof(std::size_t*));
```

Comment:
Where is the 4 coming from?
I think it should be `== n_rows * (sizeof(std::size_t) + sizeof(std::size_t*))`?
(this makes me wonder if there is any point trying to report any stats on essentially `sizeof(r1) + sizeof(r2) + sizeof(r3) + ...` -- probably that's a step too far...)

Reply:
Yes, technically `== n_rows * (sizeof(std::size_t) + sizeof(std::unique_ptr<std::size_t>))`, but at some point it feels like I'm reimplementing the formula, `4` should have us covered.

I don't understand the last part. I can't/don't now how to iterate over all `owning_handle`s. Sure, I can increment/decrement a global; but that's either a compilation flag or always a runtime cost. OTOH, there's exactly one owning handle per live row in a `soa`.

Reply:
For the last part, I was just thinking that -- because of the property you mention, that there are exactly `size()` owning handles in the world -- we _could_ include in the "bookkeeping overhead" something to do with the size of those handles. It was just an idle thought, and probably not worth pursuing.